### ISSUE TITLE
Deprecate?

### ISSUE THREAD

**Issue by eush77** [wooorm/mdast-util-escape#1]
Since [remark@3.0.0](https://github.com/wooorm/remark/releases/tag/3.0.0) it's no longer necessary to escape contents of text nodes in plugins, nor does `remark` have escape nodes anymore.

Even worse, `remark` does not complain on unknown node type, but rather fails silently producing wrong output:

``` js
> remark.stringify({ type: 'root', children: escape('foo *bar*') })
'foo \n\n\n\nbar\n\n\n'
```

 I believe for the same reason as

``` js
> remark.stringify({ type: 'toString' })
'[object Object]'
```

So `mdast-util-escape` should probably be deprecated.

Or should it? Maybe it is still used by some projects along with `mdast@<3.0.0` so it shouldn't be deprecated since older versions of `mdast` are not deprecated either.

In which case we could specify a compatible version range (`<3.0.0`) in peerDependencies but it won't work if some projects depend both on `mdast@<3.0.0` and `remark@>3.0.0` and `remark` is the module `mdast-util-escape` works in combination with. I'm not sure though if such projects exist.

What do you think?

**Comment by wooorm** [comment]
Yup, all for it, was planning on doing so pretty soon!

I think the best way to deprecate, in my experience, a project is to push a major version and deprecate that, that way people won’t get deprecation warnings for using the project, only when they’re in the process of updating it.

This project was pretty short-lived and not really marketed, so I think nobody really uses it, making it all less of a big problem.

Regarding unknown nodes, the following code:

``` js
var remark = require('remark');

var doc = remark.stringify({
    'type': 'root',
    'children': [
        {
            'type': 'paragraph',
            'children': [
                {
                    'type': 'text',
                    'value': 'foo'
                },
                {
                    'type': 'escape',
                    'value': '+'
                },
                {
                    'type': 'text',
                    'value': 'bar'
                }
            ]
        },
        {
            'type': 'escape',
            'value': '*'
        },
        {
            'type': 'paragraph',
            'children': [{
                'type': 'text',
                'value': 'baz'
            }]
        }
    ]
});

console.log('doc: ', [doc]);
```

Results in:

``` md
foobar



baz
```

...which I don’t think is very bad? Or, do you suggest throwing, or maybe compiling `value` (or `children`, when available)?

**Comment by eush77** [comment]
> ...which I don’t think is very bad?

Right, I forgot that immediate children of `root` ought to be block elements. The output is indeed not so bad.

> Or, do you suggest throwing, or maybe compiling `value` (or `children`, when available)?

No, it's just an unfortunate combination of events. Usually `remark` throws when it encounters an unknown node:

```
> remark.stringify({ type: 'foo' })
1:1: Missing compiler for node of type `foo`: `[object Object]`
```

But here `escape` is [set](https://github.com/wooorm/remark/blob/d1c52a3b28223e55aef9067ab725330019c87aae/lib/stringify.js#L802) on `compilerPrototype` and so it uses that as its compiler.

**Comment by wooorm** [comment]
Oh right, so a `visit` node would also render... We should probably fix that. :anguished:

Maybe a list of known nodes? Or, a list of disallowed methods used internally (`escape`, `enterLink`, `visit`, &c.)

And then, when a node is unknown, compile its `children` or `value`?

**Comment by eush77** [comment]
> Maybe a list of known nodes?

What if a plugin wants to extends the syntax in some way by adding tokenizers and compilers for new node types? Whitelisting existing node types would break this use case. (Unless, of course, plugins could add to this list.)

> Or, a list of disallowed methods used internally (`escape`, `enterLink`, `visit`, &c.)

Maybe we could just separate all these methods into stand-alone internal functions (if they are only used inside `remark` itself), or methods on another object (`remark.escape`) or under a single property (`compilerPrototype.utils.escape`) if they are meant to be used by plugins.

What about properties of `Object.prototype`? Are they needed for anything or would it be ok if `compilerPrototype` inherited to `null`?

> And then, when a node is unknown, compile its `children` or `value`?

Seems like a reasonable default, but at the same time it would hide errors when wrong node types are accidentally generated (due to typos or misconfigured inter-plugin dependencies).

**Upd**: regarding `Object.prototype` properties, we could just compare against them explicitly:

``` js
if (Object.prototype[type] == compiler[type]) {
   // throw
}
```

**Upd2**: Parser has `blockTokenizers` and `inlineTokenizers`, so another option is to move not utilities but rather compilers into a dedicated object (`compilerPrototype.compilers.inlineCode`).

**Comment by wooorm** [comment]
Great suggestions, `compiler.util` or externalised would be good.

However, how about the `compile` method and `type: 'compile'` nodes?

Maybe the simplest is to use an object at `compiler#compilers`, as that would allow every type of node, even `options`, `file`, `visit`, etc., and even `toString` if we use a `hasOwnProperty` check?

**Comment by eush77** [comment]
> However, how about the `compile` method and `type: 'compile'` nodes?

You're right. I agree that moving all compilers to `compiler#compilers` is less fragile and better overall. It would also add a nice symmetry with how you extend `Parser`.

The only downside that I see is that the OO style @ben-eb used in [remark-comment-blocks](https://github.com/ben-eb/remark-comment-blocks/blob/5b6f053d1d56c68ec0a72cdf81efae7ee62ccb3f/src/index.js) will not be possible anymore, but otherwise I'm all for it!

**Comment by wooorm** [comment]
One more alternative: could `Compiler`s prototype inherit from a `CompilerPrototype`? Then, utilities like `visit` are added to the `CompilerPrototype.prototype`, while `paragraph` would be on `Compiler.prototype`.

Then, we something like the following:

``` js
var visitor = this[type];

if (visitor === CompilerPrototype.prototype[type] || typeof visitor !== 'function') {
    /* throw */
}
```

Ben’s compiler would still work, but then it won’t be possible to overwrite `visit` or `compile` (I think I’ve done that some times, but that would be possible to work around I think).

**Comment by eush77** [comment]
Let's see.
1. Will it allow for new node types? If yes, then compilers would have to somehow obtain a reference to `CompilerPrototype.prototype` to use one of these utility methods. (Because there might be a compiler for `visit` or `escape` somewhere in the chain also.) In other words, repeatedly call `Object.getPrototypeOf` until `Object.prototype`. This is not pretty.
2. > but then it won’t be possible to overwrite `visit` or `compile`
   
   Maybe we could separate compiler that `unified` uses (with methods like `compile` and `visit`) from compilers for individual nodes and have the former call the latter?
   
   Then all utilities would be available on `remark.Compiler`, but plugins would add visitors to `remark.Visitor` and at some point `remark.Compiler` calls `root` visitor with arguments `compiler` and `node`, and then each visitor uses methods on `compiler` (like `compiler.all`) to recursively visit children, but no utility methods leak to `remark.Visitor` itself?
   
   And you would be able to overwrite `compile` or `visit` by changing `remark.Compiler`.

**Comment by wooorm** [comment]
Well, so my last idea wasn’t very good it seems haha.

I don’t think I would want to split this into two constructors/classes though.
Back to the `compiler#compilers` / `compiler#visitors`, what if the constructor function of `Compiler` creates that object? Meaning, each instance has its own object, subclassing constructors could add (or `null`ify) methods on that object.

I think that would still work for Ben’s use case:

``` js
// ...

class CommentBlockCompiler extends Compiler {
    constructor(file, options) {
        super(file, options);

        this.visitors.root = function (node) {
          // ...
        };
    }
}

// ...
```

However, I’m thinking in fact that instead of overloading `visitors.root`, overloading `compile` might be better here.

Do you think this would work?

**Comment by ben-eb** [comment]
I'm not sure why implementing a generic `node` visitor wouldn't work for this?

**Comment by wooorm** [comment]
What are you exactly suggesting, @ben-eb? How would that work in practise?

**Comment by eush77** [comment]
@wooorm 

> I think that would still work for Ben’s use case

I doubt that. Correct me if I'm wrong, but that approach was all about specifying node compilers as separate class methods, here one would need to put all that in `constructor`:

``` js
class CommentBlockCompiler extends Compiler {
    constructor(file, options) {
        super(file, options);

        this.visitors.root = function (node) {
          // ...
        };

        this.visitors.image = function (node) {
          // ...
        };

        this.visitors.imageReference = function (node) {
          // ...
        };
    }
}
```

And more importantly, what's the point of initializing visitors in `constructor`? What are the benefits?

Ben's approach would work if `visitors` were initialized by a constructor function right before compilation. Then plugins could subclass the constructor and override methods:

``` js
class CommentBlockCompilerVisitors extends Compiler.Visitors {
    root (node) {
        // ...
    }

    image (node) {
        // ...
    }

    imageReference (node) {
        // ...
    }
}

remark.Compiler.Visitors = CommentBlockCompilerVisitors;
```

But this is exactly as confusing as the last idea of mine.

**Comment by eush77** [comment]
> However, I’m thinking in fact that instead of overloading `visitors.root`, overloading `compile` might be better here.

I don't think I understand. Could you elaborate?

**Comment by wooorm** [comment]
@eush77 Well, if you want markdown in a JS comment, overloading `compile` to wrap in a comment would work even if someone passed a `paragraph` node, whereas right now the node _must_ be `root` to be wrapped in a comment. Then again, it won’t matter in practice, as all trees produced by `remark` are `root`

**Comment by ben-eb** [comment]
A generic node visitor could throw an exception rather than silently failing, forcing stringifiers to implement methods for types that they should be interested in. So basically any unknown node situation should not happen.

``` js
class Compiler {
  node (node) {
    throw new Error('node is not implemented.');
  }
}
```

**Comment by eush77** [comment]
@wooorm Oh, I see. I was more thinking about the general case.

@ben-eb This is actually how it's handled now.

```
> remark.stringify({ type: 'foo' })
1:1: Missing compiler for node of type `foo`: `[object Object]`
```

Or do you mean that compilers should implement _all_ nodes by themselves rather than delegate some (or most) to the basic compiler?

**Comment by ben-eb** [comment]
Ah okay, I hadn't tested that case.

**Comment by wooorm** [comment]
Okay, I’m not sure what a good solution is. :anguished: I’m moving this discussion over to remark’s repo though, hold on!

**Comment by wooorm** [comment]
Closed by 1a90999 (I like this SHA).
